substrate/client/authority-discovery/src/worker.rs

```diff
@@ -306,30 +332,50 @@ where
 	fn addresses_to_publish(&self) -> impl Iterator<Item = Multiaddr> {
 		let peer_id: Multihash = self.network.local_peer_id().into();
 		let publish_non_global_ips = self.publish_non_global_ips;
-		let addresses = self.network.external_addresses().into_iter().filter(move |a| {
-			if publish_non_global_ips {
-				return true
-			}
+		let addresses: Vec<_> = {
+			let addresses = match self.public_addresses_only {
+				Some(ref public_addresses) => public_addresses.clone(),
+				None => self
+					.network
+					.external_addresses()
+					.into_iter()
+					.filter(move |a| {
+						if publish_non_global_ips {
+							return true
+						}
+
+						a.iter().all(|p| match p {
+							// The `ip_network` library is used because its `is_global()` method is
+							// stable, while `is_global()` in the standard library currently isn't.
+							multiaddr::Protocol::Ip4(ip) if !IpNetwork::from(ip).is_global() =>
+								false,
+							multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() =>
+								false,
+							_ => true,
+						})
+					})
+					.collect(),
+			};
 
-			a.iter().all(|p| match p {
-				// The `ip_network` library is used because its `is_global()` method is stable,
-				// while `is_global()` in the standard library currently isn't.
-				multiaddr::Protocol::Ip4(ip) if !IpNetwork::from(ip).is_global() => false,
-				multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() => false,
-				_ => true,
-			})
-		});
+			// The address must include the peer id if not already set.
+			addresses
+				.into_iter()
+				.map(move |a| {
+					if a.iter().any(|p| matches!(p, multiaddr::Protocol::P2p(_))) {
+						a
+					} else {
+						a.with(multiaddr::Protocol::P2p(peer_id))
+					}
+				})
+				.collect()
+		};
 
-		debug!(target: LOG_TARGET, "Authority DHT record peer_id='{:?}' addresses='{:?}'", peer_id, addresses.clone().collect::<Vec<_>>());
+		debug!(
```

Comment:
dq:
I think moving the debug log here changes a bit the details we expose.
I was curious to know the addresses we publish in DHT prior to adding the `/p2p/local_peer_id`.

Having the log line here might hint us about `/p2p/other_peer` entries (because they will remain untouched).
Could you see any value in distinguishing between `rogue_ip_address` and `rogue_ip_address/p2p/local_peer_id`?

Reply:
I moved the log here to see what addresses exactly are published to DHT. In case we want pretty addresses without `/p2p/...` part, we would also need to drop the `/p2p/...` part if it was provided by the operator.